op-verifier/contracts/OPVerifier.sol

```diff
@@ -1,43 +1,177 @@
+    function getL2Output(
+        uint256 _l2OutputIndex
+    ) external view returns (Types.OutputProposal memory);
+}
+
+interface IOptimismPortal {
+    function l2Oracle() external view returns (IL2OutputOracle);
+
+    function disputeGameFactory() external view returns (IDisputeGameFactory);
+
+    function respectedGameType() external view returns (GameType);
 }
 
 contract OPVerifier is IEVMVerifier {
-    error OutputRootMismatch(uint256 l2OutputIndex, bytes32 expected, bytes32 actual);
+    error OutputRootMismatch(
+        OPWitnessProofType proofType,
+        uint256 index,
+        bytes32 expected,
+        bytes32 actual
+    );
+    error GameTypeMismatch(
+        uint256 disputeGameIndex,
+        GameType expected,
+        GameType actual
+    );
+    error GameChallenged(uint256 disputeGameIndex);
+    error GameTooEarly(
+        uint256 disputeGameIndex,
+        uint256 age,
+        uint256 minimumAge
+    );
+    error DeprecatedProof(
+        OPWitnessProofType proofType,
+        uint256 disputeGameIndex
+    );
+
+    IOptimismPortal public immutable optimismPortal;
+    uint256 public immutable disputeGameMinAge;
 
-    IL2OutputOracle public opOracle;
     string[] _gatewayURLs;
 
-    constructor(string[] memory urls, address outputOracle) {
+    constructor(
+        string[] memory urls,
+        address optimismPortalAddress,
+        uint256 minimumAge
+    ) {
         _gatewayURLs = urls;
-        opOracle = IL2OutputOracle(outputOracle);
+        optimismPortal = IOptimismPortal(optimismPortalAddress);
+        disputeGameMinAge = minimumAge;
     }
 
-    function gatewayURLs() external view returns(string[] memory) {
+    function gatewayURLs() external view returns (string[] memory) {
         return _gatewayURLs;
     }
 
-    function getStorageValues(address target, bytes32[] memory commands, bytes[] memory constants, bytes memory proof) external view returns(bytes[] memory values) {
-        (OPWitnessData memory opData, StateProof memory stateProof) = abi.decode(proof, (OPWitnessData, StateProof));
-        Types.OutputProposal memory l2out = opOracle.getL2Output(opData.l2OutputIndex);
-        bytes32 expectedRoot = Hashing.hashOutputRootProof(opData.outputRootProof);
-        if(l2out.outputRoot != expectedRoot) {
-            revert OutputRootMismatch(opData.l2OutputIndex, expectedRoot, l2out.outputRoot);
+    function getL2OracleOutput(uint256 index) internal view returns (bytes32) {
+        IL2OutputOracle opOracle = optimismPortal.l2Oracle();
+
+        // If opOracle address is zero then revert with deprecated
+        if (address(opOracle) == address(0))
+            revert DeprecatedProof(OPWitnessProofType.L2OutputOracle, index);
+
+        // Get L2 Output Proposal at index
+        Types.OutputProposal memory l2out = opOracle.getL2Output(index);
+        return l2out.outputRoot;
+    }
+
+    function getDisputeGameOutput(
```

Comment:
Are there any reference you used so that I can verify?

Reply:
Would be `proveWithdrawalTransaction`

https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L254-L328

Reply:
How did you manage to test each failure scenario like  GameChallenged, GameTooEarly, DeprecatedProof or are they even possible to test?